File: CppOptimizer/ConsoleApplication/ConsoleApplication/MathC.cpp
#include "MathC.h"

#include <numeric>

static bool PointsWithinTriangle2D(const std::vector<double>& point, const std::vector<double>& a,
                                   const std::vector<double>& b,
                                   const std::vector<double>& c, const double tolerance = .001f)
{
    const double w1 = (a.at(0) * (c.at(1) - a.at(1)) + (point.at(1) - a.at(1)) * (c.at(0) - a.at(0)) - point.at(0) * (c.
            at(1) -
            a.at(1))) /
        ((b.at(1) - a.at(1)) * (c.at(0) - a.at(0)) - (b.at(0) - a.at(0)) * (c.at(1) - a.at(1)));

    const double w2 = (point.at(1) - a.at(1) - w1 * (b.at(1) - a.at(1))) /
        (c.at(1) - a.at(1));

    return w1 >= tolerance && w2 >= tolerance && w1 + w2 <= 1.0 - tolerance;
}

static bool LineIntersect2D(const std::vector<double>& start1, const std::vector<double>& end1,
                            const std::vector<double>& start2,
                            const std::vector<double>& end2)
{
    //Line1
    const double a1 = end1.at(1) - start1.at(1);
    const double b1 = start1.at(0) - end1.at(0);
    const double c1 = a1 * start1.at(0) + b1 * start1.at(1);

    //Line2
    const double a2 = end2.at(1) - start2.at(1);
    const double b2 = start2.at(0) - end2.at(0);
    const double c2 = a2 * start2.at(0) + b2 * start2.at(1);

    const double denominator = a1 * b2 - a2 * b1;

    if (denominator == 0.0)
        return false;

    const std::vector<double> point = {(b2 * c1 - b1 * c2) / denominator, (a1 * c2 - a2 * c1) / denominator};

    if (point == start1 || point == end1 ||
        point == start2 || point == end2)
        return false;

    constexpr double tolerance = .001f;

    return point.at(0) > std::min(start1.at(0), end1.at(0)) + tolerance &&
        point.at(0) < std::max(start1.at(0), end1.at(0)) - tolerance &&
        point.at(0) > std::min(start2.at(0), end2.at(0)) + tolerance &&
        point.at(0) < std::max(start2.at(0), end2.at(0)) - tolerance &&
        point.at(1) > std::min(start1.at(1), end1.at(1)) + tolerance &&
        point.at(1) < std::max(start1.at(1), end1.at(1)) - tolerance &&
        point.at(1) > std::min(start2.at(1), end2.at(1)) + tolerance &&
        point.at(1) < std::max(start2.at(1), end2.at(1)) - tolerance;
}

static bool TriangleIntersect2D(const std::vector<double>& a1, const std::vector<double>& a2,
                                const std::vector<double>& a3,
                                const std::vector<double>& b1, const std::vector<double>& b2,
                                const std::vector<double>& b3)
{
    return LineIntersect2D(a1, a2, b1, b2) ||
        LineIntersect2D(a1, a3, b1, b2) ||
        LineIntersect2D(a2, a3, b1, b2) ||
        LineIntersect2D(a1, a2, b1, b3) ||
        LineIntersect2D(a1, a3, b1, b3) ||
        LineIntersect2D(a2, a3, b1, b3) ||
        LineIntersect2D(a1, a2, b2, b3) ||
        LineIntersect2D(a1, a3, b2, b3) ||
        LineIntersect2D(a2, a3, b2, b3);
}

static std::vector<double> ClosetPointOnLine(const std::vector<double>& point, const std::vector<double>& start,
                                             const std::vector<double>& end)
{
    //Get heading
    std::vector<double> heading = {end.at(0) - start.at(0), end.at(1) - start.at(1)};
    const double magnitude = sqrt(pow(heading.at(0), 2) + pow(heading.at(1), 2));
    heading = {heading.at(0) / magnitude, heading.at(1) / magnitude};

    //Do projection from the point but clamp it
    std::vector<double> lhs = {point.at(0) - start.at(0), point.at(1) - start.at(1)};
    double dotP = std::inner_product(lhs.begin(), lhs.end(), heading.begin(), 0);

    dotP = dotP < 0.0 ? 0.0 : dotP > magnitude ? magnitude : dotP;

    return {start.at(0) + heading.at(0) * dotP, start.at(1) + heading.at(1) * dotP};
}

template <typename T>
static std::list<T> TSharedBetween(const std::list<T> target, const std::list<T> other, int max = -1)
{
}

static float QuickSquareDistance(const std::vector<double>& point1, const std::vector<double>& point2)
{
    return pow(point1.at(0) - point2.at(0), 2) + pow(point1.at(1) - point2.at(1), 2);
}

File: UnityProjectForGeneration/JsonGenerator/Assets/Scripts/MathC.cs
#region Libraries

using System;
using UnityEngine;

#endregion

public static class MathC
{
    public static bool LineIntersect2DWithTolerance(Vector2 start1, Vector2 end1, Vector2 start2, Vector2 end2)
    {
        //Line1
        float a1 = end1.y - start1.y;
        float b1 = start1.x - end1.x;
        float c1 = a1 * start1.x + b1 * start1.y;

        //Line2
        float a2 = end2.y - start2.y;
        float b2 = start2.x - end2.x;
        float c2 = a2 * start2.x + b2 * start2.y;

        float denominator = a1 * b2 - a2 * b1;

        if (denominator == 0)
            return false;

        Vector2 point = new Vector2((b2 * c1 - b1 * c2) / denominator, (a1 * c2 - a2 * c1) / denominator);

        if (point == start1 || point == end1 ||
            point == start2 || point == end2)
            return false;

        const float tolerance = .001f;

        return point.x > MathF.Min(start1.x, end1.x) + tolerance &&
               point.x < MathF.Max(start1.x, end1.x) - tolerance &&
               point.x > MathF.Min(start2.x, end2.x) + tolerance &&
               point.x < MathF.Max(start2.x, end2.x) - tolerance &&
               point.y > MathF.Min(start1.y, end1.y) + tolerance &&
               point.y < MathF.Max(start1.y, end1.y) - tolerance &&
               point.y > MathF.Min(start2.y, end2.y) + tolerance &&
               point.y < MathF.Max(start2.y, end2.y) - tolerance;
    }

    public static bool PointWithinTriangle2DWithTolerance(Vector2 point, Vector2 a, Vector2 b, Vector2 c,
        float tolerance = .001f)
    {
        float s1 = c.y - a.y + 0.0001f;
        float s2 = c.x - a.x;
        float s3 = b.y - a.y;
        float s4 = point.y - a.y;

        float w1 = (a.x * s1 + s4 * s2 - point.x * s1) / (s3 * s2 - (b.x - a.x + 0.0001f) * s1);
        float w2 = (s4 - w1 * s3) / s1;
        return w1 >= tolerance && w2 >= tolerance && w1 + w2 <= 1f - tolerance;
    }
    
    public static bool TriangleIntersect2D(Vector2 a1, Vector2 a2, Vector2 a3, Vector2 b1, Vector2 b2, Vector2 b3)
    {
        return LineIntersect2DWithTolerance(a1, a2, b1, b2) ||
               LineIntersect2DWithTolerance(a1, a3, b1, b2) ||
               LineIntersect2DWithTolerance(a2, a3, b1, b2) ||
               LineIntersect2DWithTolerance(a1, a2, b1, b3) ||
               LineIntersect2DWithTolerance(a1, a3, b1, b3) ||
               LineIntersect2DWithTolerance(a2, a3, b1, b3) ||
               LineIntersect2DWithTolerance(a1, a2, b2, b3) ||
               LineIntersect2DWithTolerance(a1, a3, b2, b3) ||
               LineIntersect2DWithTolerance(a2, a3, b2, b3);
    }

    public static Vector2 ClosetPointOnLine(Vector2 point, Vector2 start, Vector2 end)
    {
        //Get heading
        Vector2 heading = end - start;
        float magnitudeMax = heading.magnitude;
        heading.Normalize();

        //Do projection from the point but clamp it
        Vector2 lhs = point - start;
        float dotP = Vector2.Dot(lhs, heading);
        dotP = Mathf.Clamp(dotP, 0f, magnitudeMax);

        return start + heading * dotP;
    }
}